### PR TITLE
Store User-Agent as a Part of Review Statistics

### LINKED ISSUE #697: Store User-Agent in Review Statistics
**As a** Zally administrator/user
**I want** to know which reviews have been made by which client type
**So that** I can categorize the linting requests

**Motivation**
It allows us to understand our users better. We can at least learn which client has been used by the user: directly (e.g. `curl`), `go-cli`, `web-ui`, etc.

**Acceptance Criteria**
- [ ] [`User-Agent`](https://tools.ietf.org/html/rfc7231#section-5.5.3) is stored as a part of linting result
- [ ] `ReviewStatistic` in `/review-statistics` endpoint contains this information
- [ ] `/review-statistics` endpoint has a `user-agent` filter

### PR DESCRIPTION
**Pull request by maxim-tschumak**
Closes #697.

### CODE REVIEW

File: server/src/test/java/de/zalando/zally/statistic/RestReviewStatisticsTest.java [hunk 1]
```diff
@@ -10,6 +10,7 @@
 import de.zalando.zally.rule.zalando.ZalandoRuleSet;
 import de.zalando.zally.util.ErrorResponse;
 import de.zalando.zally.util.TestDateUtil;
+import org.junit.Ignore;
```
[the remark this answers is not in the document]
Comment (netme) [reply server/src/test/java/de/zalando/zally/statistic/RestReviewStatisticsTest.java]: Yes, please :-)

File: server/src/main/java/de/zalando/zally/statistic/ReviewStatisticsController.java [hunk 2]
```diff
@@ -46,14 +47,28 @@ public ReviewStatistics retrieveReviewStatistics(
...
 
-        LOG.info("Found {} api reviews from {} to {}", apiReviews.size(), from, to);
+        LOG.info("Found {} api reviews from {} to {} user_agent {}", apiReviews.size(), from, to, userAgent);
         return new ReviewStatistics(apiReviews);
     }
 
+    private Collection<ApiReview> findReviews(LocalDate from, LocalDate to, String userAgent) {
+        Collection<ApiReview> apiReviews;
+        LocalDate t = to != null ? to : today();
+        if (userAgent == null || userAgent.length() == 0) {
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/statistic/ReviewStatisticsController.java]:
```java
if (userAgent == null || userAgent.isEmpty() {
...
```

File: server/src/main/java/de/zalando/zally/statistic/ReviewStatisticsController.java [hunk 3]
```diff
@@ -46,14 +47,28 @@ public ReviewStatistics retrieveReviewStatistics(
...
+            apiReviews = from != null
+                ? apiReviewRepository.findByDayBetween(from, t)
+                : apiReviewRepository.findAllFromLastWeek();
+        } else {
+            apiReviews = from != null
+                ? apiReviewRepository.findByUserAgentAndDayBetween(userAgent, from, t)
+                : apiReviewRepository.findByUserAgentFromLastWeek(userAgent);
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/statistic/ReviewStatisticsController.java]: Should we have a smarter `findBy...` functions which will allow empty `userAgent` values? Then we can skip this `if` statement.

File: server/src/main/java/de/zalando/zally/statistic/ApiReviewStatistic.java [hunk 4]
```diff
@@ -54,4 +56,12 @@ public void setNumberOfEndpoints(int numberOfEndpoints) {
     public void setViolations(List<RuleViolation> violations) {
         this.violations = violations;
     }
+
+    public void setUserAgent(String userAgent) {
+        this.userAgent = userAgent;
+    }
+
+    public String getUserAgent() {
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/statistic/ApiReviewStatistic.java]: Minor: I will swap positions of getter and setter.

File: server/src/main/java/de/zalando/zally/apireview/ApiViolationsController.java [hunk 5]
```diff
@@ -58,11 +58,9 @@ public ApiDefinitionResponse validate(@RequestBody ApiDefinitionRequest request,
         RulesPolicy requestPolicy = retrieveRulesPolicy(request);
 
         List<Result> violations = rulesValidator.validate(apiDefinition, requestPolicy);
-        apiReviewRepository.save(new ApiReview(request, apiDefinition, violations));
+        apiReviewRepository.save(new ApiReview(request, userAgent, apiDefinition, violations));
 
-        ApiDefinitionResponse response = buildApiDefinitionResponse(violations, userAgent);
-
-        return response;
+        return buildApiDefinitionResponse(violations, userAgent);
```
Comment (netme) [inline server/src/main/java/de/zalando/zally/apireview/ApiViolationsController.java]: Great

**Comment by maxim-tschumak** [comment]
Thanks for your comments @netme

**Comment by maxim-tschumak** [comment]
:+1:

**Comment by maxim-tschumak** [comment]
:+1:

**Comment by tkrop** [comment]
:+1:

**Pull request merged**